server/pubsub/pub.go

```diff
@@ -19,71 +19,48 @@ import (
 	"sync"
 )
 
-type subscriber struct {
-	receiver Receiver
+// Message defines a published message.
+type Message struct {
+	// ID identifies this message.
+	ID string `json:"id,omitempty"`
+
+	// Data is the actual data in the entry.
+	Data []byte `json:"data"`
+
+	// Labels represents the key-value pairs the entry is labeled with.
+	Labels map[string]string `json:"labels,omitempty"`
 }
 
-type publisher struct {
+// Receiver receives published messages.
+type Receiver func(Message)
+
+type Publisher struct {
 	sync.Mutex
 
-	topics map[string]*topic
+	subs map[*Receiver]bool
```

Comment:
Why does it needs to be bool? I don't see where false value would be used
If map value is not used than `struct{}` is better option

Reply:
struct is faster and uses less memory than bool: https://itnext.io/set-in-go-map-bool-and-map-struct-performance-comparison-5315b4b107b

Reply:
is still valid also in 1.21:
```
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
                           │  bench.txt  │
                           │   sec/op    │
SetWithBoolValueWrite        3.154 ± 10%
SetWithStructValueWrite      3.075 ± 10%
SetWithInterfaceValueWrite   5.760 ±  5%

                           │  bench.txt   │
                           │     B/op     │
SetWithBoolValueWrite        811.4Mi ± 0%
SetWithStructValueWrite      737.6Mi ± 0%
SetWithInterfaceValueWrite   1.800Gi ± 0%

                           │  bench.txt  │
                           │  allocs/op  │
SetWithBoolValueWrite        380.2k ± 0%
SetWithStructValueWrite      379.8k ± 0%
SetWithInterfaceValueWrite   379.6k ± 0%
```

Reply:
Thanks for sharing ❤️